state/operation.go

```diff
@@ -0,0 +1,266 @@
+func (op *operation) Id() string {
+	return op.st.localID(op.doc.DocId)
+}
+
+// Tag implements the Entity interface and returns a names.Tag that
+// is a names.ActionTag.
+func (op *operation) Tag() names.Tag {
+	return op.OperationTag()
+}
+
+// OperationTag returns the operation's tag.
+func (op *operation) OperationTag() names.OperationTag {
+	return names.NewOperationTag(op.Id())
+}
+
+// Enqueued returns the time the action was added to state as a pending
+// Action.
+func (op *operation) Enqueued() time.Time {
+	return op.doc.Enqueued
+}
+
+// Started returns the time that the Action execution began.
+func (op *operation) Started() time.Time {
+	return op.doc.Started
+}
+
+// Completed returns the completion time of the Operation.
+func (op *operation) Completed() time.Time {
+	return op.doc.Completed
+}
+
+// Summary is the reason for running the operation.
+func (op *operation) Summary() string {
+	return op.doc.Summary
+}
+
+// Status returns the final state of the operation.
+// If not explicitly set, this is derived from the
+// status of the associated actions/tasks.
+func (op *operation) Status() ActionStatus {
+	if op.doc.Status != ActionPending {
+		return op.doc.Status
+	}
+	statusStats := set.NewStrings()
+	for _, s := range op.taskStatus {
+		statusStats.Add(string(s))
+	}
+	for _, s := range statusOrder {
+		if statusStats.Contains(string(s)) {
+			return s
+		}
+	}
+	return op.doc.Status
+}
+
+// Refresh refreshes the contents of the operation.
+func (op *operation) Refresh() error {
+	doc, taskStatus, err := op.st.getOperationDoc(op.Id())
+	if err != nil {
+		if errors.IsNotFound(err) {
+			return err
+		}
+		return errors.Annotatef(err, "cannot refresh operation %v", op.Id())
+	}
+	op.doc = *doc
+	op.taskStatus = taskStatus
+	return nil
+}
+
+var statusOrder = []ActionStatus{
+	ActionRunning,
+	ActionPending,
+	ActionFailed,
+	ActionCancelled,
+	ActionCompleted,
+}
+
+var statusCompletedOrder = []ActionStatus{
+	ActionFailed,
+	ActionCancelled,
+	ActionCompleted,
+}
+
+// newAction builds an Action for the given State and actionDoc.
+func newOperation(st *State, doc operationDoc, taskStatus []ActionStatus) Operation {
+	return &operation{
+		st:         st,
+		doc:        doc,
+		taskStatus: taskStatus,
+	}
+}
+
+// newOperationDoc builds a new operationDoc.
+func newOperationDoc(mb modelBackend, summary string) (operationDoc, string, error) {
+	id, err := sequenceWithMin(mb, "task", 1)
+	if err != nil {
+		return operationDoc{}, "", errors.Trace(err)
+	}
+	operationId := strconv.Itoa(id)
+	modelUUID := mb.modelUUID()
+	return operationDoc{
+		DocId:     mb.docID(operationId),
+		ModelUUID: modelUUID,
+		Enqueued:  mb.nowToTheSecond(),
+		Status:    ActionPending,
+		Summary:   summary,
+	}, operationId, nil
+}
+
+// EnqueueOperation records the start of an operation.
+func (m *Model) EnqueueOperation(summary string) (string, error) {
+	doc, id, err := newOperationDoc(m.st, summary)
```

Comment:
Shouldn't this be in a transaction builder due to the monotonic key ids?

Reply:
yes it should, good catch